TypeSystem/types.h

```diff
@@ -44,12 +44,28 @@ struct CurrentSuper {
 // The superclass for all Current structures.
 struct CurrentStruct : CurrentSuper {};
 
-#define IS_CURRENT_STRUCT(T) (std::is_base_of<::current::CurrentStruct, ::current::decay<T>>::value)
-
 // The superclass for `Variant` type.
 struct CurrentVariant : CurrentSuper {};
 
+template<typename NAME>
+struct CurrentVariantImpl : CurrentVariant {
+  static const char* VariantName() {
+    return NAME::VariantNameImpl();
+  }
+};
+
+struct CurrentVariantDefaultName final {
+  static const char* VariantNameImpl() {
+    return "Variant";
+  }
+};
+
+#define IS_CURRENT_STRUCT(T) (std::is_base_of<::current::CurrentStruct, ::current::decay<T>>::value)
 #define IS_VARIANT(T) (std::is_base_of<::current::CurrentVariant, ::current::decay<T>>::value)
+#define IS_CURRENT_STRUCT_OR_VARIANT(T) (std::is_base_of<::current::CurrentSuper, ::current::decay<T>>::value)
+
+#define IS_NONEMPTY_CURRENT_STRUCT(T) \
```

Comment:
The reason was indeed convoluted. Look, this:
```
if (IS_CURRENT_STRUCT(X) && !IS_NONEMPTY_CURRENT_STRUCT(X)) {
```
is now this:
```
if (IS_EMPTY_CURRENT_STRUCT(X)) {
```

Two good news are: we have tests, and they passed. :+1: bbfc94a